plugin-interface/src/main/scala/mesosphere/marathon/plugin/auth/AuthorizedAction.scala

```diff
@@ -9,7 +9,7 @@ import mesosphere.marathon.plugin.{ Group, RunSpec }
   *
   * @tparam R the type of the resource.
   */
-sealed trait AuthorizedAction[+R]
+sealed trait AuthorizedAction[-R]
```

Comment:
Why do we want that?

This means given `A <: B` that `AuthorizedAction[A] :> AuthorizedAction[B]`. So If `A` is a sub resource of `B` any action of `A` is allowed on `B` but not any action on `B` is allowed on `A`. Did I get this correct?

Reply:
The biggest issue is that we saw a lot of code that was failing (Ex: https://github.com/dcos/metronome/pull/161/files) 
The issue with that code is `case Some(run) => request.authorized(ViewRunSpec, run, Ok(run))` passed `run` which was of type `StartedJobRun`... which is incorrect.  When I investigated... any value was possible to pass in here... because with `[+R]` the `AuthorizationAction[A]` will look for an `A` type in the superclass hierarchy...  AND type inferencing... it went all the way up to `Any`.      We had a 1 place in metronome and several across several versions of Marathon where the wrong type of `R` was passed in for the AuthorizedAction type.   `-R` will allow that type or a subtype.  This is useful in that a compile time failure will occur if the wrong type is used (we had several places looking for ViewRunSpec auth or a Group action which is obviously incorrect).

Reply:
Yeah, Covariance definitely seems wrong. An `AuthorizedAction[RunSpec]` can be used with a `Group` (RunSpec and Group both implement Any). The likely intention was the other way: an `AuthorizedAction[RunSpec]` can be used for anything that implements `RunSpec` (which the contra-variance allows)

Relevant is the type spec for the method used here:

```
authorizer.isAuthorized(identity, UpdateRunSpec, app)
// isAuthorized[Resource](Identity, AuthorizedAction[Resource], Resource)
```

(hopefully this helps shed light on why covariance is wrong)

Reply:
Not sure I get it. I understand the type inference but I'd rather start at the types themselves. Just changing the type hierarchy because the type inference was wrong seems not the right reason to change this as one can always help the inference by providing types explicitly.